preview/interfaces/QrCode.hpp

```diff
@@ -0,0 +1,307 @@
+            {
+                uint8_t numBlocks = numErrorCorrectionBlocks[static_cast<uint8_t>(ecc)][version - 1];
+                uint16_t totalEcc = numErrorCorrectionCodewords[static_cast<uint8_t>(ecc)][version - 1];
+                return totalEcc / numBlocks;
+            }
+
+            static constexpr uint8_t ModeBitsNumeric(uint8_t version)
+            {
+                if (version <= 9)
+                    return 10;
+                if (version <= 26)
+                    return 12;
+                return 14;
+            }
+
+            static constexpr uint8_t ModeBitsAlphanumeric(uint8_t version)
+            {
+                if (version <= 9)
+                    return 9;
+                if (version <= 26)
+                    return 11;
+                return 13;
+            }
+
+            static constexpr uint8_t ModeBitsLatin1(uint8_t version)
+            {
+                if (version <= 9)
+                    return 8;
+                if (version <= 26)
+                    return 16;
+                return 16;
+            }
+
+        public:
+            static constexpr std::size_t MaxSizeNumeric(uint8_t version, QrCodeEcc ecc)
+            {
+                uint16_t moduleCount = numRawDataModulesForVersion[version];
+                uint16_t dataCapacityInBytes = moduleCount / 8 - numErrorCorrectionCodewords[static_cast<uint8_t>(ecc)][version - 1];
+                uint16_t digitBits = dataCapacityInBytes * 8 - 4 - ModeBitsNumeric(version);
+
+                uint16_t blocksOf3 = digitBits / 10;
+                digitBits -= blocksOf3 * 10;
+
+                return blocksOf3 * 3 + (digitBits >= 7 ? 2 : digitBits >= 4 ? 1
+                                                                            : 0);
+            }
+
+            static constexpr std::size_t MaxSizeAlphanumeric(uint8_t version, QrCodeEcc ecc)
+            {
+                uint16_t moduleCount = numRawDataModulesForVersion[version];
+                uint16_t dataCapacityInBytes = moduleCount / 8 - numErrorCorrectionCodewords[static_cast<uint8_t>(ecc)][version - 1];
+                uint16_t digitBits = dataCapacityInBytes * 8 - 4 - ModeBitsAlphanumeric(version);
+
+                uint16_t blocksOf2 = digitBits / 11;
+                digitBits -= blocksOf2 * 11;
+
+                return blocksOf2 * 2 + digitBits / 6;
+            }
+
+            static constexpr std::size_t MaxSizeLatin1(uint8_t version, QrCodeEcc ecc)
+            {
+                uint16_t moduleCount = numRawDataModulesForVersion[version];
+                uint16_t dataCapacityInBytes = moduleCount / 8 - numErrorCorrectionCodewords[static_cast<uint8_t>(ecc)][version - 1];
+                uint16_t digitBits = dataCapacityInBytes * 8 - 4 - ModeBitsLatin1(version);
+
+                return digitBits / 8;
+            }
+
+        public:
+            template<uint8_t Version, QrCodeEcc ecc>
```

Comment:
Why is the Version a template parameter? You already have it in the constructor

Reply:
This is a nested struct, where the version and ecc are needed compile-time to determine storage

Reply:
Ohhh right! Ok in that case it makes sense as a template parameter. But then you can already define the version as constexpr to be used in other places?